### PR TITLE
[mlir][linalg] Enable fusion by expansion of reduction and named ops

### PR DESCRIPTION
**Pull request by qedawkins**
This adds support for expansion of named linalg ops and linalg ops with reduction iterators. This improves the ability to make fusion decisions WRT reduction operations. To recover the previous behavior, users of the patterns can add a control function to restrict propagation of reshape by expansion through linalg ops with reduction iterators.

For named linalg ops, this always converts the named op into a generic.

### CODE REVIEW

File: mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp [hunk 1]
```diff
@@ -848,6 +853,11 @@ fuseWithReshapeByExpansion(GenericOp genericOp, Operation *reshapeOp,
   // The iterator types of the expanded op are all parallel.
   SmallVector<utils::IteratorType> iteratorTypes(
       expansionInfo.getExpandedOpNumDims(), utils::IteratorType::parallel);
+  for (auto [i, type] : llvm::enumerate(genericOp.getIteratorTypesArray())) {
+    ReassociationIndicesRef group = expansionInfo.getExpandedDims(i);
+    for (auto i : group)
+      iteratorTypes[i] = type;
```
Comment (hanhanW) [inline mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:859]: nit1: should replace the second `i` variable with name `j`.
nit2: the group is only used here, how about directly pass it to for op? I.e.,

```
for (auto j : expansionInfo.getExpandedDims(i))
  iteratorTypes[j] = type;
```

**Review by hanhanW — COMMENTED** [review]
I don't really follow these logics these days, so I suggest to wait for Mahesh.

**Comment by qedawkins** [comment]
sounds good, thanks for taking a look!

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
This is OK to land, but you can change the utility method here to work on any `LinalgOp`. I am suspecting for.your use case you are generalizing the op before applying the reshape. We can skip that step.

**Comment by qedawkins** [comment]
> This is OK to land, but you can change the utility method here to work on any `LinalgOp`. I am suspecting for.your use case you are generalizing the op before applying the reshape. We can skip that step.

Done. The same should probably be done for the collapsing patterns, but I assume that's fine to do as a follow up?

**Review by MaheshRavishankar — APPROVED** [review]
Nice ! Thanks!

**Pull request merged**